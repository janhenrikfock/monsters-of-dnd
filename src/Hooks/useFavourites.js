import { useState } from 'react'

export default function useFavourites(monsterDetails) {
  const [favourites, setFavourites] = useState([])

  function addFavourite(index) {
    const previous = monsterDetails.slice(0, index)
    const marked = monsterDetails.splice(index, 1)
    const following = monsterDetails.slice(index + 1, -1)

    const newFavourite = {
      ...marked,
      favourite: true,
    }
    setFavourites([...previous, newFavourite, ...following])
  }

  return { favourites, addFavourite }
}
