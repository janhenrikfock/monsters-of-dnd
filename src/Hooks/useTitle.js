import { useEffect } from 'react'

export function useTitle(monster) {
  useEffect(() => {
    if (monster) {
      document.title = monster.name
    } else {
      document.title = 'Monsters of DnD'
    }
  })
}
