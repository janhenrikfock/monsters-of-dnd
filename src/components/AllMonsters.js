import React, { useEffect, useState } from 'react'
import styled from 'styled-components/macro'
import saveLocally from './lib/saveLocally'
import OneMonsterItem from './OneMonsterItem'
import LoadingAnimation from './LoadingAnimation'

export default function AllMonsters() {
  const [monsterDetails, setMonsterDetails] = useState([])

  const [loading, setLoading] = useState(false)

  useEffect(() => {
    const locallyLoadedMonsterNames = JSON.parse(
      localStorage.getItem('monsternames')
    )

    if (!locallyLoadedMonsterNames || locallyLoadedMonsterNames.length === 0) {
      setLoading(true)
      fetch(`https://www.dnd5eapi.co/api/monsters`)
        .then((res) => res.json())
        .then((data) => {
          saveLocally('monsternames', data.results)
          data.results.forEach((monster) => {
            fetch(`https://www.dnd5eapi.co${monster.url}`)
              .then((res) => res.json())
              .then((data) => {
                setMonsterDetails((monsterDetails) => [...monsterDetails, data])
              })
          })
        })
        .then(setLoading(false))
        .catch((err) => console.log(err))
    }
  }, [])

  if (monsterDetails.length !== 0) {
    saveLocally('monsterdetails', monsterDetails)
  } else if (localStorage.getItem('monsterdetails')) {
    setMonsterDetails(JSON.parse(localStorage.getItem('monsterdetails')))
  }

  if (loading) {
    return <LoadingAnimation />
  } else {
    return (
      <>
        <HeadlineStyled>Monsters of DnD</HeadlineStyled>
        {monsterDetails.map((monster, index) => (
          <OneMonsterItem key={index} monster={monster}></OneMonsterItem>
        ))}
      </>
    )
  }
}

const HeadlineStyled = styled.h1`
  font-size: 150%;
  text-align: center;
  border-bottom: 8px solid #f50808;
`
