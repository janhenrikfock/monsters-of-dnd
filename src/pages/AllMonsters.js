import React, { useState } from 'react'
import styled from 'styled-components/macro'
import LoadingAnimation from './LoadingAnimation'
import OneMonsterItem from '../components/OneMonsterItem'
import Fuse from 'fuse.js'

export default function AllMonsters({ loading, monsterDetails }) {
  const [searchResults, setSearchResults] = useState('')

  const fuse = new Fuse(monsterDetails, {
    keys: ['name', 'type', 'challenge_rating'],
  })
  const results = searchResults ? fuse.search(searchResults) : monsterDetails

  if (loading) {
    return <LoadingAnimation />
  } else {
    return (
      <>
        <Header>
          <Headline>Monsters of DnD</Headline>
          <Input
            type="text"
            value={searchResults}
            onChange={handleInput}
            placeholder={'Search for Monster'}
          />
        </Header>
        <Main>
          {results.map((monster) => (
            <OneMonsterItem key={monster.index} monster={monster} />
          ))}
        </Main>
      </>
    )
  }
  function handleInput({ currentTarget = {} }) {
    const { value } = currentTarget
    setSearchResults(value)
  }
}
const Header = styled.header`
  position: fixed;
  top: 0px;
  border-bottom: 1px solid #d0d0cd;
  padding-bottom: 10px;
  width: 100%;
  background: white;
`
const Main = styled.main`
  margin-top: 120px;
`
const Headline = styled.h1`
  color: black;
  font-size: 150%;
  text-align: center;
  border-bottom: 3px solid var(--highlightcolor);
  font-family: serif;
  padding-bottom: 13px;
`
const Input = styled.input`
  width: 80%;
  line-height: 26px;
  margin-left: 10%;
  display: block;
  text-align: center;
  font-size: 18px;
`
