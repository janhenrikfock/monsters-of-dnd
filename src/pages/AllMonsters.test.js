import React from 'react'
import { MemoryRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { render } from '@testing-library/react'
import { useTitle } from '../Hooks/useTitle'

import AllMonsters from './AllMonsters'

describe('AllMonsters', () => {
  useTitle = useTitle()
  const monsterDetails = [
    {
      name: 'Monstername',
      type: 'type',
      challenge_rating: '00',
      index: 'monstername',
    },
  ]

  it('shows headline Monsters of DnD', () => {
    const { getByText } = render(
      <Router>
        <AllMonsters monsterDetails={monsterDetails} useTitle={useTitle} />
      </Router>
    )
    expect(getByText('Monsters of DnD')).toBeInTheDocument()
  })

  it('renders OneMonsterItem correctly', () => {
    const list = renderer.create(
      <Router>
        <AllMonsters monsterDetails={monsterDetails} />
      </Router>
    )
    expect(list).toMatchSnapshot()
  })
})
