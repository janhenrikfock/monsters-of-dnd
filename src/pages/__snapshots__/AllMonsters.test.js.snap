// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllMonsters renders OneMonsterItem correctly 1`] = `
Array [
  <header
    className="AllMonsters__Header-sc-8qeyzf-0 iTyJER"
  >
    <h1
      className="AllMonsters__Headline-sc-8qeyzf-2 hElxVk"
    >
      Monsters of DnD
    </h1>
    <input
      className="AllMonsters__Input-sc-8qeyzf-3 cUVJaY"
      onChange={[Function]}
      placeholder="Search for Monster"
      type="text"
      value=""
    />
  </header>,
  <main
    className="AllMonsters__Main-sc-8qeyzf-1 cfHHgp"
  >
    <div
      className="sc-AxjAm dLSYMf"
    >
      <div
        className="sc-AxirZ CVuNr"
      >
        <h2
          className="sc-AxiKw cUhBKK"
        >
          Monstername
        </h2>
        <div
          className="sc-AxhCb ceVZWe"
        >
          <div
            className="sc-AxhUy gWaiz"
          >
            <p
              className="sc-AxgMl iKoGXN"
            >
              type
            </p>
            <p
              className="sc-AxheI bQSAct"
            />
          </div>
          <p
            className="sc-Axmtr VxVIM"
          >
            CR:
            00
          </p>
        </div>
      </div>
      <a
        aria-current={null}
        href="/monster/monstername"
        onClick={[Function]}
      >
        <img
          alt="view details"
          className="sc-AxmLO kLvddQ"
          src="book.png"
        />
      </a>
    </div>
  </main>,
]
`;
